File: server/socket.js
module.exports = (io) => {
    io.on('connection', (socket) => {
        socket.on('join-room', (room) => {
            socket.join(room);
            socket.on('greeting-message', (data) => {
                io.to(room).emit('greeting-message', {
                    message: data.message,
                    author: data.author
                });
            });
            socket.on('chat-message', (data) => {
                io.to(room).emit('chat-message', {
                    message: data.message,
                    author: data.author
                });
            });
        });
    });
};

File: internal/socket/socket.go
package socket

import (
	"encoding/json"

	socketio "github.com/googollee/go-socket.io"
)

type ChatMsg struct {
	Message		string	`json:"message"`
	Author		string	`json:"author"`
}

func Socket(server *socketio.Server) {
	server.On("connection", func(so socketio.Socket) {
		so.On("join-room", func(roomID string) {
			so.Join(roomID)
	
			so.On("greeting-message", func(msg ChatMsg) {
				msgJSON, _ := json.Marshal(msg)
				server.BroadcastTo(roomID, "greeting-message", string(msgJSON))
			})
	
			so.On("chat-message", func(msg ChatMsg) {
				msgJSON, _ := json.Marshal(msg)
				server.BroadcastTo(roomID, "chat-message", string(msgJSON))
			})
		})
	})
}
